### ISSUE TITLE
Passing "Accuracy" to model.compile() parameter "metrics" (instead of "accuracy") returns 0 / epoch without throwing an error

### ISSUE THREAD

**Issue by Akshaysehgal2005** [tensorflow/tensorflow#46436]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS Big Sur 11.0.1
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version: Python 3.8.3
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Current Behavior**
It seems that when "Accuracy" is passed as a string (instead of "accuracy" lowercase) to the metric parameter, the compile deserializes it to get  `tf.keras.metrics.Accuracy` directly instead of fetching `tf.keras.metrics.MeanMetricWrapper` as defined in `_get_metric_object()` from compile_utils. Therefore, the accuracy metric is still visible during Epochs but incorrectly calculated  (always equal to 0) without throwing any warnings/errors.

**Expected Behavior**
Passing "Accuracy" instead of "accuracy" should result in a ValueError instead of failing silently and returning 0 value for each epoch. 

**Reproducable code**
For scenario 1 (correct case)- 

```python
#Scenario 1: Lower case accuracy (accuracy displayed is proper)

from tensorflow import keras
from tensorflow.keras import layers

X_train = np.random.random((100,8))
y_train = np.random.randint(0,2,(100,))

model = keras.Sequential(
    [
     layers.Dense(10,activation="relu",input_shape=(8,)),
     layers.Dense(10,activation="relu"),
     layers.Dense(1,activation="sigmoid")
    ]
)

model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
model.fit(X_train,y_train,batch_size=64,epochs=2,verbose=2)
print(model.metrics)

# Epoch 1/2
# 2/2 - 0s - loss: 0.6900 - accuracy: 0.4900
# Epoch 2/2
# 2/2 - 0s - loss: 0.6892 - accuracy: 0.5000   #<------
        
# [<tensorflow.python.keras.metrics.Mean object at 0x7f90e6626d90>, 
#  <tensorflow.python.keras.metrics.MeanMetricWrapper object at 0x7f90e5ac73a0>]  #<------
```

And scenario 2 (Incorrect case)- 

```python
#Scenario 2: Uppercase case Accuracy (accuracy displayed / epoch always 0)

from tensorflow import keras
from tensorflow.keras import layers

X_train = np.random.random((100,8))
y_train = np.random.randint(0,2,(100,))

model = keras.Sequential(
    [
     layers.Dense(10,activation="relu",input_shape=(8,)),
     layers.Dense(10,activation="relu"),
     layers.Dense(1,activation="sigmoid")
    ]
)

model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['Accuracy'])
model.fit(X_train,y_train,batch_size=64,epochs=2,verbose=2)
print(model.metrics)

# Epoch 1/2
# 2/2 - 0s - loss: 0.7021 - accuracy: 0.0000e+00
# Epoch 2/2
# 2/2 - 0s - loss: 0.6999 - accuracy: 0.0000e+00   #<------
        
# [<tensorflow.python.keras.metrics.Mean object at 0x7f90c80e4bb0>, 
#  <tensorflow.python.keras.metrics.Accuracy object at 0x7f90ca039a30>]  #<------
```

I don't think this was intentional, else it should have thrown a ValueError. I have described the details on this SO post.

**Comment by Akshaysehgal2005** [comment]
I traced the GitHub implementation for compile find out the reason. Here are the steps as per my understanding followed by the reason. (Code line links inline)

1. During compile, when a metric is passed, the parameter is stored in a [MetricsContainer][2] object.
2. This container class then calls a function called [_get_metric_object][3] from `compile_utils.py`. The job of this function is to take the input and return a metric object of that metric's class.
3. One of the first things this function does is to [check][4] if the input belongs to list `['accuracy', 'acc', 'crossentropy', 'ce']` or not.
    - If `YES`, then it directly fetches the classes from the metrics.py and [calls][5] the `MeanMetricWrapper` class. The [job of this class][6] is to wrap a stateless metric function with the Mean metric. This calculates the mean of the metric you have added.
    - If `NO`, then it calls the [get function][7] from metrics.py. The `get` function further calls a [deserialize function][8] whose job is call a function call [deserialize_keras_object][9] function from `utils.generic_utils.py`. This function's job is to take the string and retrieve the actual object!

Let's see the 2 scenarios now.

### Scenario 1: "accuracy"

```
#With lower case accuracy
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])
print(model.metrics)
```
```
[<tensorflow.python.keras.metrics.Mean at 0x7f90c7ea10d0>,
 <tensorflow.python.keras.metrics.MeanMetricWrapper at 0x7f90c7d07e20>]
```

Since the metric provided belongs to the `['accuracy', 'acc', 'crossentropy', 'ce']`, the `_get_metric_object` function fetches the `tf.keras.metrics.Accuracy` class and explicitly passes it to the `tf.keras.metrics.MeanMetricWrapper`. This calculates the mean accuracy as expected.

### Scenario 2: "Accuracy"

```
#With upper case Accuracy
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['Accuracy'])
print(model.metrics)
```
```
[<tensorflow.python.keras.metrics.Mean at 0x7f90e7285e20>,
 <tensorflow.python.keras.metrics.Accuracy at 0x7f90e72fceb0>]
```

Since the "Accuracy" doesnt belong to the list, `_get_metric_object` calls the `metrics.get()` ->`metrics.deserialize()`-> `generic_utils.deserialize_keras_object()` function which simply pulls up the `tf.keras.metrics.Accuracy` and returns that directly, instead of calling `tf.keras.metrics.MeanMetricWrapper`.

This is why you get incorrect values for accuracy, but it does not throw an error

**Comment by ravikyram** [comment]
I have tried in colab with TF 2.3, 2.4 gist [here](https://colab.research.google.com/gist/ravikyram/bed7c74d6a29378fe3947a08af3255f2/untitled619.ipynb) anf TF nightly version(`2.5.0-dev20210114` gist [here](https://colab.research.google.com/gist/ravikyram/8f76a13cf5ba44de0224993ca5d57722/untitled620.ipynb#scrollTo=hnEoHTmJbh02) and was able to reproduce the issue. Thanks!

**Comment by Akshaysehgal2005** [comment]
Hi, I have made a pull request with a potential solution to this. https://github.com/tensorflow/tensorflow/pull/46441. Do check. Thanks!

**Comment by Akshaysehgal2005** [comment]
Hi, is anything needed from my side?

**Comment by Akshaysehgal2005** [comment]
@jvishnuvardhan any update on this? is anything needed from my side? thanks!

**Comment by sachinprasadhs** [comment]
Hi, Now it produces accuracy for both `metrics=['accuracy'] `and `metrics=['Accuracy']`, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/7e1091663e53f045096817b81461eab1/untitled619.ipynb#scrollTo=DkmS3e3uZOuc) and conform the same. Thanks!
